File: client/src/features/auth/api/registration.ts
import axios from 'axios'

interface RegistrationData {
  email_entered: boolean
  otp_verified: boolean
  password_is_set: boolean
}

const instance = axios.create({
  withCredentials: true,
  baseURL: `${import.meta.env.VITE_API_BASE}/registration`
})

const registration = (email: string) => instance.post('', { email })

const resendRegistrationOtp = () => instance.post('/resend_otp')

const verifyEmail = (otp: string) => instance.post('/verify_email', { otp })

const setRegistrationPassword = (password: string) => instance.post('/set_password', { password })

const completeRegistration = (password: string) => instance.post('/complete', { password })

const fetchRegistrationData = () => instance.get<Record<string, RegistrationData>>('/data')

export {
  completeRegistration,
  fetchRegistrationData,
  registration,
  resendRegistrationOtp,
  setRegistrationPassword,
  verifyEmail
}

File: server/app/api/routes/registration.py
from app.core.errors import INTERNAL_SERVER_ERROR
from app.db.database import get_db
from app.schemas.registration import *
from app.services.registration_service import *
from fastapi import APIRouter, Depends, HTTPException, Response

router = APIRouter()


@router.post("/registration")
def registration(
    data: RegistrationRequest, response: Response, db: Session = Depends(get_db)
):
    try:
        return start_registration(data.email, response, db)
    except HTTPException as e:
        raise e
    except Exception as e:
        print(e)
        raise HTTPException(status_code=500, detail=INTERNAL_SERVER_ERROR)


@router.get("/registration/data")
async def data(
    request: Request,
    response: Response,
    db: Session = Depends(get_db),
):
    try:
        return await registration_data(request, response, db)
    except HTTPException as e:
        raise e
    except Exception as e:
        print(e)
        raise HTTPException(status_code=500, detail=INTERNAL_SERVER_ERROR)


@router.post("/registration/resend_otp")
async def resend(
    request: Request,
    response: Response,
    db: Session = Depends(get_db),
):
    try:
        return await resend_otp(request, response, db)
    except HTTPException as e:
        raise e
    except Exception as e:
        print(e)
        raise HTTPException(status_code=500, detail=INTERNAL_SERVER_ERROR)


@router.post("/registration/verify_email")
async def verify(
    data: OTPVerificationRequest,
    request: Request,
    response: Response,
    db: Session = Depends(get_db),
):
    try:
        return await verify_email(request, response, db)
    except HTTPException as e:
        raise e
    except Exception as e:
        print(e)
        raise HTTPException(status_code=500, detail=INTERNAL_SERVER_ERROR)


@router.post("/registration/set_password")
async def password(
    data: PasswordRequest,
    request: Request,
    response: Response,
    db: Session = Depends(get_db),
):
    try:
        return await set_password(request, response, db)
    except HTTPException as e:
        raise e
    except Exception as e:
        print(e)
        raise HTTPException(status_code=500, detail=INTERNAL_SERVER_ERROR)


@router.post("/registration/complete")
async def complete(
    data: PasswordRequest,
    request: Request,
    response: Response,
    db: Session = Depends(get_db),
):
    try:
        return await complete_registration(request, response, db)
    except HTTPException as e:
        raise e
    except Exception as e:
        print(e)
        raise HTTPException(status_code=500, detail=INTERNAL_SERVER_ERROR)
